test/benchmark/avro_upsert/generator/generate_upserts.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import argparse
+import pathlib
+import subprocess
+
+SHARED_FILES = "/usr/share/generator"
+
+
+def run(args: argparse.Namespace) -> None:
+    """Run the generator, inserting args.num_messages number of messages."""
+
+    messages_per_process = int(args.num_messages / args.parallelism)
+
+    value_schema = pathlib.Path(SHARED_FILES, "value-schema.json").read_text().strip()
+    value_distribution = (
+        pathlib.Path(SHARED_FILES, "value-distribution.json").read_text().strip()
+    )
+
+    kafka_gen = [
+        "/usr/local/bin/kgen",
+        "--quiet",
+        "--bootstrap-server",
+        "kafka:9092",
+        "--schema-registry",
+        "http://schema-registry:8081",
+        "--num-messages",
+        str(messages_per_process),
+        "--topic",
+        "upsertavrotest",
+        "--partitions",
+        "30",
+        "--keys",
+        "random",
+        "--key-min",
+        "0",
+        "--key-max",
+        str(args.num_keys),
+        "--values",
+        "avro",
+        "--avro-schema",
+        value_schema,
+        "--avro-distribution",
+        value_distribution,
+    ]
+
+    print(
+        f"Spawning {args.parallelism} generator processes, writing {messages_per_process} messages each"
+    )
+    procs = [subprocess.Popen(kafka_gen) for _ in range(0, args.parallelism)]
+    for (i, p) in enumerate(procs):
+        p.wait()
+        print(
+            f"{i}/{args.parallelism} processes finished: pid={p.pid} returncode={p.returncode}"
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-k",
+        "--num-keys",
+        type=int,
+        default=25000000,
+        help="Number of unique keys to generate (should match value-distribution!)",
+    )
+    parser.add_argument(
+        "-n",
+        "--num-messages",
+        type=int,
+        default=400000000,
+        help="Total number of messages to generate",
+    )
+    parser.add_argument(
+        "-p", "--parallelism", type=int, default=40, help="Number of processes to spawn"
+    )
+
+    args = parser.parse_args()
+    run(args)
```

Comment:
I'm surprised that wrapping the Rust in the Python was required for performance! Like, I'd expect the Rust to just be fast enough to saturate the network. Should be super easy to spawn a little thread pool with Rayon if required.

Anyway, 100% not worth changing around right this second, but food for thought for the future!

Reply:
IME, generating all these random values and then populating the avro object becomes the bottleneck before a good network connection does.

Reply:
I can look at using Rayon -- I was only using Python as a replacement for `xargs` or `parallel`, as well as reading the schema / distribution files send those into `kgen`.

I'll make a new PR to use `rayon` . What do we think of allowing the `schema` / `distribution` arguments to be filenames?

Reply:
I think it's fine to script the parallelism in Python for now (also talked about this with Nikhil a few mins ago). If we ever start doing this more often though, we can add parallelism to the script itself using Rayon or similar.

Reply:
Sounds good. I'll go ahead and merge this as is!